src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1361,15 +1434,31 @@ public SolutionCompilationState WithFrozenSourceGeneratedDocuments(
                     // Just compute the checksum from the source text passed in.
                     originalSourceTextChecksum: null,
                     generationDateTime);
-                documentStates.Add(newGeneratedState);
+
+                if (syntaxNode != null)
+                {
+                    newGeneratedState = newGeneratedState.WithSyntaxRoot(syntaxNode);
+                }
+
+                documentStates.Add(newGeneratedState.Id, newGeneratedState);
+            }
+        }
+
+        if (FrozenSourceGeneratedDocumentStates is not null)
+        {
+            // We also carry forward any previously frozen source generated documents that were not part of this freeze. This ensures multiple
+            // freezes are additive, and everything will be unfrozen correctly, without us needing to have a chain of compilation trackers
+            foreach (var (id, state) in FrozenSourceGeneratedDocumentStates.States)
+            {
+                documentStates.TryAdd(id, state);
             }
         }
 
         // If every document we looked at matched what we've already generated, we have nothing new to do
         if (documentStates.Count == 0)
             return this;
 
-        var documentStatesByProjectId = documentStates.ToDictionary(static state => state.Id.ProjectId);
+        var documentStatesByProjectId = documentStates.Values.ToMultiDictionary(static state => state.Id.ProjectId);
```

Comment:
why a multi dict?

Reply:
Not sure if the name of the method is correct, but this returns a `Dictionary<key, ImmutableArray<value>>`. Its the same as what the old `ToDictionary` was doing when `documentStates` was an immutable array.